tokio/src/io/async_fd.rs

```diff
@@ -508,6 +508,92 @@ impl<T: AsRawFd> AsyncFd<T> {
     pub async fn writable_mut<'a>(&'a mut self) -> io::Result<AsyncFdReadyMutGuard<'a, T>> {
         self.readiness_mut(Interest::WRITABLE).await
     }
+
+    /// Reads or writes from the file descriptor using a user-provided IO operation.
+    ///
+    /// This method is useful to retry an operation that might block (by returning
+    /// a [`WouldBlock`] error) until it doesn't block any more.
+    ///
+    /// The readiness of the file descriptor is awaited and when the file descriptor is ready,
+    /// the provided closure is called. The closure should attempt to perform
+    /// IO operation on the file descriptor by manually calling the appropriate syscall.
+    /// If the operation fails because the file descriptor is not actually ready,
+    /// then the closure should return a [`WouldBlock`] error. In such case the
+    /// readiness flag is cleared and the file descriptor readiness is awaited again.
+    /// This loop is repeated until the closure returns an `Ok` or an error
+    /// other than [`WouldBlock`].
```

Comment:
Reading through the documentation for `AsyncFd`, I think this would be useful:
```suggestion
    /// The `async_io` method is a convenience utility that waits for the file
    /// descriptor to become ready, and then executes the provided IO operation.
    /// Since file descriptors may be marked ready spuriously, the closure will
    /// be called repeatedly until it returns something other than a
    /// `WouldBlock` error. This is done using the following loop:
    ///
    /// ```no_run
    /// # use std::io::{self, Result};
    /// # struct Dox<T> { inner: T }
    /// # impl<T> Dox<T> {
    /// #     async fn writable(&self) -> Result<&Self> {
    /// #         Ok(self)
    /// #     }
    /// #     fn try_io<R>(&self, _: impl FnMut(&T) -> Result<R>) -> Result<Result<R>> {
    /// #         panic!()
    /// #     }
    /// async fn async_io<R>(&self, mut f: impl FnMut(&T) -> io::Result<R>) -> io::Result<R> {
    ///     loop {
    ///         // or `readable` if called with the read interest.
    ///         let guard = self.writable().await?;
    ///
    ///         match guard.try_io(&mut f) {
    ///             Ok(result) => return result,
    ///             Err(_would_block) => continue,
    ///         }
    ///     }
    /// }
    /// # }
    /// ```
```

Reply:
I added this (and didn't know about that `Dox` trick, that's neat!)